### PR TITLE
Handle unencrypted message in subscribe with crypto

### PR DESCRIPTION
**Pull request by wkal-pubnub**

### CODE REVIEW

File: src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt [hunk 1]
```diff
@@ -201,34 +204,56 @@ internal class SubscribeMessageProcessor(
...
+        val inputText = if (pubnub.mapper.isJsonObject(input)) {
+            // property pn_other is used when we want to send encrypted Push Notification, not whole JSON object is encrypted but only value of pn_other property
+            if (pubnub.mapper.hasField(input, PN_OTHER)) {
+                // JSON with pn_other property indicates that this is encrypted Push Notification
+                pubnub.mapper.elementToString(input, PN_OTHER)
+            } else {
+                // plain JSON object indicates that this is not encrypted message
+                return input to logAndReturnDecryptionError()
+            }
+        } else if (input.isJsonPrimitive && input.asJsonPrimitive?.isString == true) {
```
Comment (marcin-cebo) [inline src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt:229]: This condition doesn't exist in JsonElement.processHistoryMessage.
What do you think about adding it there ? or maybe we could use JsonElement.processHistoryMessage method here to reduce code duplication?

File: src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt [hunk 2]
```diff
@@ -201,34 +204,56 @@ internal class SubscribeMessageProcessor(
...
+            // String may represent not encrypted string or encrypted data. We will check this when decrypting.
             pubnub.mapper.elementToString(input)
```
Comment (marcin-cebo) [inline src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt:231]: If we decide not to use JsonElement.processHistoryMessage here. I would advise to write unit tests for this method similar to those from JsonElement.processHistoryMessage plus extra test to handles:
"else if (input.isJsonPrimitive && input.asJsonPrimitive?.isString == true)"

File: src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt (hunk 1 above)
Comment (wkal-pubnub) [reply src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt:229]: you're right, I thought maybe these functions differ somehow, but they are actually exactly the same (or should be)
I'll reuse it

File: src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt (hunk 2 above)
Comment (wkal-pubnub) [reply src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt]: this case is tested already both in JsonElement test and SubscribeMessageProcessorTest (all cases - including strings, arrays, objects, pn_other)

File: src/test/kotlin/com/pubnub/extension/JsonElementTest.kt [hunk 3]
```diff
@@ -1,67 +1,76 @@
...
         fun cryptoModuleConfiguration(): List<Arguments> {
             val cipherKey = "enigma"
             return listOf(
                 Arguments.of(
-                    "Hello world.",
+                    JsonPrimitive("Hello world."),
                     false,
                     "X/wLZqR/I+4eU7WwtKycTwy9+L9e7m2BhklK9ZQWtwI=",
                     CryptoModule.createLegacyCryptoModule(cipherKey = cipherKey, randomIv = true),
                     generatePayload()
```
Comment (marcin-cebo) [inline src/test/kotlin/com/pubnub/extension/JsonElementTest.kt:30]: I guess we can remove 'generatePayload()' from all Arguments.

**Review by marcin-cebo — APPROVED** [review]

File: src/test/kotlin/com/pubnub/extension/JsonElementTest.kt (hunk 3 above)
Comment (wkal-pubnub) [reply src/test/kotlin/com/pubnub/extension/JsonElementTest.kt]: Done

**Pull request merged**